feat(curriculum): add business card lab

zairahira opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

As this is the first CSS lab, the user stories are not open-ended in terms of design.
I have asked campers to add specific colours, margins, and borders.

To do:
- [ ] Add demo project link.

<!-- Feel free to add any additional description of changes below this line -->

zairahira committed:
feat/add-business-card-lab

zairahira committed:
feat: add tests

zairahira committed:
fix: update project description

zairahira committed:
fix failing test

moT01 reviewed (commented):
In the HTML lab I added, I used the format `lab-recipe-page` for the block dashed name and filenames - it doesn't sound as good, but it keeps all the labs next to each other in the file tree - the URL will be `lab-recipe-page` though, so there's some trade-offs. The display name is still set in the intro.json file, so we can show whatever we want.

I think we should be consistent, so we should either change this to use that format, or I can change the other one. What do you think @zairahira?

zairahira committed:
fix: organise project

Dario-DC commented on client/i18n/locales/english/intro.json:
```diff
@@ -1699,6 +1699,12 @@
+      "lab-business-card": {
+        "title": "Design a Business Card",
+        "intro": [
+          "In this lab, you'll create a business card and style it using CSS"
```
```suggestion
          "In this lab, you'll create a business card and style it using CSS."
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
```diff
@@ -0,0 +1,468 @@
+
+# --description--
+
+**Objective:** Design a Business card that looks similar to <a href="https://codepen.io/rancid_beet/full/ExBaJdw" target="_blank" rel="noopener noreferrer nofollow">https://business-card.freecodecamp.rocks</a>. Fulfill the user stories below and get all the tests to pass to complete the lab. 
```
```suggestion
**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. 
```
To keep the same format in all labs.

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(new __helpers.CSSHelp(document).getStyle('p').getPropVal('margin'), '5px 0px');
+```
+
+The links of the page should have no underline by default 
````
```suggestion
The links of the page should have no underline by default.
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.match(document.querySelector('.profile-image').getAttribute('src'),/.+/);
+```
+
+The `alt` attribute of the image should be set to a meaningful description.
````
It always pass.

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.exists(document.querySelector('div.business-card p.company'));
+```
+
+The first `p` element with the class `full-name` should contain your name.
````
It always pass.

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(new __helpers.CSSHelp(document).getStyle('.full-name').getPropVal('color'), 'darkslategray');
+```
+
+The second `p` element with the class `designation` should contain your designation.
````
It always pass

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(new __helpers.CSSHelp(document).getStyle('.designation').getPropVal('color'), 'gray');
+```
+
+The third `p` element with the class `company` should contain your company name.
````
It always pass

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(new __helpers.CSSHelp(document).getStyle('hr').getPropVal('margin'), '20px 0px');
+```
+
+After the first `hr` element, there should be a `p` element with an email address as your text.
````
It always pass

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(document.querySelector('.business-card a').textContent, 'Portfolio');
+```
+
+Your portfolio `a` element should have a valid `href`.
````
It always pass

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(document.querySelectorAll(".social-media a")[0].textContent, 'Twitter');
+```
+
+The Twitter `href` should point to a valid link.
````
It always pass

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(document.querySelectorAll(".social-media a")[1].textContent, 'LinkedIn');
+```
+
+The LinkedIn `href` should point to a valid link.
````
It always pass

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,468 @@
+assert.equal(document.querySelectorAll(".social-media a")[2].textContent, 'GitHub');
+```
+
+The GitHub `href` should point to a valid link.
````
It always pass

Dario-DC reviewed (commented):
I just had a brief look. There are several tests that pass without adding any code.

zairahira committed:
fix: apply suggestions from code review

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

moT01 commented on curriculum/challenges/_meta/lab-business-card/meta.json:
```diff
@@ -0,0 +1,11 @@
+  "isUpcomingChange": true,
+  "usesMultifileEditor": true,
+  "dashedName": "lab-business-card",
+  "order": 2,
```
The issue about the order should be fixed, so you should be able to change this to whatever.

moT01 replied:
Also, I updated the spreadsheet to fill in the block order - it looks like this one is 38

zairahira committed:
marking tests

zairahira committed:
fix tests

zairahira committed:
cleanup

zairahira committed:
fix: project order

zairahira committed:
Merge upstream

zairahira committed:
fix: improve tests

zairahira committed:
fix: add html boilerplate

zairahira committed:
nits

moT01 reviewed (changes requested):
I was able to follow the instructions and pass all the tests, so I think I would be fine with this 👍 I do have a number of suggestions. ..

Since this is the first CSS project, we could maybe add a user story to "link the styles.css file". I don't know if it's needed, but I couldn't figure out why my styles weren't being applied for a bit. Maybe users will be used to it at this point.

I personally think the user stories and tests could be trimmed down quite a bit, and allow for more flexibility. For instance, many of the elements ask for a specific background color - similarly with font-size and other properties. This makes it so many user projects will end up looking very similar - I would prefer to have some more flexibility. Maybe just one or two where you give something specific and then allow anything in other areas - or not even include a user story if it's been covered. For instance, you could say "The body should have a background color of rosybrown." and then maybe "other thing should have a background color set to something other than the default." Or with the font-size, just have them set a font-size for one or two things, and leave out the others - and they can set it to anything if they want to or just leave it. The goal of implementing font-size would be accomplished.

I don't think that stuff needs to be added here, I'm just explaining how I think it could be better - kind of just a preference.

All that said, I did have fun doing it and it covers what it needs to 👍

zairahira committed:
fix: add user story for linking styles.css

zairahira committed:
fix: trim user stories and tests

zairahira committed:
fix: firefox tests

zairahira committed:
improve vague tests

zairahira committed:
fix: apply suggestions from code review

zairahira commented:
Thank you @moT01 for a detailed review. 

I have made the following changes:

- Added a story and tests for linking style sheet to HTML.
- Trimmed the user stories and adjusted the tests accordingly.
- Fixed the failing test for Firefox.
- Improved vague instructions/ tests.

zairahira committed:
Merge branch 'main' into feat/add-first-css-lab

ojeytonwilliams committed:
fix: add closing brackets

zairahira committed:
fix according to naming conventions

zairahira committed:
add block type

zairahira committed:
Merge remote-tracking branch 'upstream/main' into feat/add-first-css-lab

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
```diff
@@ -0,0 +1,418 @@
+---
+id: 6690e10ebe2181212abc9652
+title: Design a Business Card
```
```suggestion
title: Build a Business Card
```

moT01 replied:
Do we want to strictly use "Build a (project)"? I kinda liked "Design" here, but I'm maybe we want to just stick with "Build" for consistency.

jdwilkin4 replied:
I was thinking about this and here are my two cents

For all HTML, JS, React /Typescript, and Backend labs it should use the Build naming convention.
For select CSS labs, where appropriate, it is fine if it is Design X naming convention

Dario-DC commented on curriculum/challenges/_meta/lab-business-card/meta.json:
```diff
@@ -0,0 +1,12 @@
+  "dashedName": "lab-business-card",
+  "order": 38,
+  "superBlock": "front-end-development",
+  "challengeOrder": [{ "id": "6690e10ebe2181212abc9652", "title": "Design a Business Card" }],
```
```suggestion
  "challengeOrder": [{ "id": "6690e10ebe2181212abc9652", "title": "Build a Business Card" }],
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,418 @@
+assert.exists(document.querySelector('img.profile-image'));
+```
+
+The image source should be set to a valid image.
````
Once `img.profile-image`  is created this is always satisfied, even if you don't set the `src`.
I would check the length of the attribute instead.

zairahira replied:
It should work now.

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,418 @@
+assert.isAbove($('img')[0].alt.length, 0);
+```
+
+Inside the `.business-card` class, after the `img` element, there should be a `p` element with the class `full-name`.
````
```suggestion
Inside the `.business-card` element, after the `img` element, there should be a `p` element with the class `full-name`.
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,418 @@
+assert.exists(document.querySelector('div.business-card div.social-media'));
+```
+
+Inside the `.social-media` class, there should be an `h2` element with the text `Connect with me`.
````
```suggestion
Inside the `.social-media` element, there should be an `h2` element with the text `Connect with me`.
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,418 @@
+assert.equal(document.querySelector('.social-media h2').textContent, 'Connect with me');
+```
+
+Inside the `.social-media` class, there should be an `a` element with the text `Twitter` 
````
```suggestion
Inside the `.social-media` element, there should be an `a` element with the text `Twitter`.
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,418 @@
+assert.isAbove($('.social-media a')[0].getAttribute('href').length, 0);
+```
+
+Inside the `.social-media` class, there should be an `a` element with the text `LinkedIn`.
````
```suggestion
Inside the `.social-media` element, there should be an `a` element with the text `LinkedIn`.
```

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
````diff
@@ -0,0 +1,418 @@
+assert.isAbove($('.social-media a')[1].getAttribute('href').length, 0);
+```
+
+Inside the `.social-media` class, there should be an `a` element with the text `GitHub`
````
```suggestion
Inside the `.social-media` element, there should be an `a` element with the text `GitHub`.
```

Dario-DC reviewed (commented):
The project is looking good :+1:  I left a few comments.

Since this is about styling, I would also check that the image size is appropriate. Set its width should be enough.

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
```diff
@@ -0,0 +1,418 @@
+1. All links on the page should have a default of no underline. When you hover over the links, there should be an underline.
+1. You should have a `div` with the class `business-card`. 
+1. The `business-card` class should be `300px` wide and it should have any background color of your choice. There should be a padding of `20px` all around and a top margin of `100px`. The text should be center aligned and the font size should be `16px`. The left and right margins should be set to `auto`.
+1. Inside the `buisness-card` div, there should be an `img` element with the class `profile-image`. You can set the image source to `https://cdn.freecodecamp.org/curriculum/labs/flower.jpg` if you like. There should be an `alt` with a meaningful description.
```
```suggestion
1. Inside the `business-card` div, there should be an `img` element with the class `profile-image`. You can set the image source to `https://cdn.freecodecamp.org/curriculum/labs/flower.jpg` if you like. There should be an `alt` with a meaningful description.
```

zairahira committed:
add demoType

zairahira committed:
fix: apply suggestions from code review

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>
Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>

zairahira committed:
fix: update test

zairahira committed:
merge main

zairahira committed:
fix: update intro.json

ojeytonwilliams commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
```diff
@@ -0,0 +1,419 @@
+1. You should link the style sheet `styles.css` using the `link` tag in the `head` section of the HTML file.
+1. Your page background should be `rosybrown`. The overall font should be `Arial`, with a fallback of `sans-serif`.   
+1. All of the `p` elements should have a top and bottom margin of `5px`.
+1. All links on the page should have a default of no underline. When you hover over the links, there should be an underline.
```
```suggestion
1. By default, all links on the page should not be underlined. When you hover over the links, they should be underlined.
```
The first sentence implied (to me) that "no underline" was a property that the links could have.

ojeytonwilliams reviewed (approved):
Hey @zairahira, since I'll be away for a bit and don't want to be in the way, I've left an approval.

I did notice one minor thing, though, but it's not blocking, imo.

jdwilkin4 committed:
Update curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md

Co-authored-by: Oliver Eyton-Williams <ojeytonwilliams@gmail.com>

jdwilkin4 reviewed (approved):
LGTM 🚀

zairahira committed:
update title and add do not copy

Dario-DC commented on curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md:
```diff
@@ -0,0 +1,419 @@
+
+# --description--
+
+**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab. **Do not copy this demo project**
```
```suggestion
**Objective:** Fulfill the user stories below and get all the tests to pass to complete the lab.
```
As I understand we are not going to add that in the description of each project.

moT01 replied:
I'm going to commit this so we can get this merged. I don't think any of the other projects have it. We can go through and make sure they are consistent later as well.

moT01 committed:
Update curriculum/challenges/english/25-front-end-development/lab-business-card/6690e10ebe2181212abc9652.md

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>
